### ISSUE TITLE
network plugin swallowing samples?

### ISSUE THREAD

**Issue by mrunge** [collectd/collectd#2873]
*   Version of collectd: 5.8
*   Operating system / distribution: CentOS

## Actual behavior
We were using collectd-tg to create samples, like `collectd-tg -n 100 -p 1 -H 1 -i 300 -d dest_host -D 25826`
On the receivers side _dest_host_, there is the network plugin runing, and also the csv plugin configured to write out samples:

```
LoadPlugin csv

<Plugin csv>
    DataDir "/collectd/data/"
</Plugin>
```

Now, after sending the first batch of 100 samples, ctrl-c is pressed, to stop collectd-tg. Inspecting the output directory for collectd never shows the desired number of 100 samples/files. I usually get to 40 or 50 out of expected 100 samples. When using larger numbers, like 10k, you'll usually get 9970 (or so). 
If you wait a few intervals, the amount _n_ of values is reached, but not all csv files contain the same
number of samples.

For tests, make sure to stop collectd, clean out /collectd/data and start collectd again.
I've also tested with write_prometheus instead of write_csv, and did not reach the expected amount of metrics.

I'd be expecting some buffers to get filled and flushed when full; however, they should get flushed at some point anyways.

Any idea, what's causing this?

**Comment by faxm0dem** [comment]
Remember the network plugin uses UDP. So at least two things can happen:

1. The datagrams reach dest_host not chronologically
2. Some datagrams get lost

You can easily test for 1. by searching the log for relevant messages. As for two, you could check the values of the `protocols` plugin on dest_host.

**Comment by mrunge** [comment]
Right.
However in this test it doesn't matter, if datagrams are not getting in in chronological order, as long as they get in.
UDP is not that reliable, that is clear. I'd expect to loose one or the other package, but not ~ 50% when sending 100 samples.

**Comment by faxm0dem** [comment]
It does matter because collectd will throw the late events away

**Comment by mrunge** [comment]
`collectd-tg -n 100 -H 1 -p 100 -i 300 -d dest_host` creates in this test 63 directories,
some with 3 values, some with 2, some with 1, in total 96 files.

I would have expected 100 directories, all with 1 value.

**Comment by faxm0dem** [comment]
The documentation states:

```
       -n num_vl
           Sets the number of unique value lists (VL) to generate. Defaults to 10000.
       -p num_plugins
           Sets the number of unique plugins to simulate. Defaults to 20.
```

I admit I'm not sure what to expect for instance when `-n = -p`

I conducted a few experiments using collectd-tg (sending to localhost) and unixsock plugin, and while the results are a bit whacky with low values of `-n` they seem pretty consistent. 

Using your cmdline (`-p 100 -n 100`) I can always see 100 values (constantly querying unixsock).

Could you try running collectd-tg on the same host as the server to rule out network effects?

**Comment by faxm0dem** [comment]
I can however confirm your finding that in the case of `-p 100 -n 100` there are not 100 unique plugins. There *are* 100 values though, so I think you're chasing a red herring here.
I still believe you're having transport issues

**Comment by mrunge** [comment]
We could reproduce this behaviour in at least 3 different environments. (all virtualized, all bare metal, mixture (bare metal and vm deployment))

**Comment by faxm0dem** [comment]
Can you reproduce it using localhost?

**Comment by mrunge** [comment]
Yes, there is some inconsistency on the same node, I can see what you described 3 comments above (not 100 unique plugins, values distributed over the plugins). But, I don't get 100 values, in my case `collectd-tg -n 100 -p 100 -H 1 -d 127.0.0.1 - i 300`, I get 49 values.

However, I assume there is a buffer involved, which might not be flushed.
